holidayextras/generator-ui-component
# Tar 197 - Extend generator to create non-shared views in apps

#### What does this PR do?
It would be nice to be able to easily build the non shared React views inside our apps (that are based on the webapp-template). To this end I've extended the generator so that it checks for a `package.json` file in any parent directory, if it finds one then it assumes that we're inside a webapp-template project and begins the process of building an internal non-shared view. If it doesn't find a `package.json` file then it begins the process of building an external shared view.

#### What unit or integration tests does this PR have?
I've rewritten the original tests as they now only apply to the external generator. I've added new tests for the master generator and the internal one.

#### What selenium tests does this PR have?
none.

#### How should a developer review this?
Check over the code to make sure I've not left any daft things in there and that the code makes sense and is legible.

#### How should this be manually tested?
After checking out this repo you will have to run `npm link` inside it to make it globally accessible to the yeoman generator. You will also have to have `yo` and `browserify` installed globally.

You may run into an error EMFILE. This is due to Browserify attempting to open too many files in parallel. Adding `ulimit -n 2560` to your bash_profile should resolve this.

1. checkout the webapp-temaplte repo and run the generator inside it using `yo ui-component` and check the internal generator was used.
  - you should be asked which feature to build for
  - you should see the new view in the feature folder chosen.
2. run the generator using `yo ui-component` in a directory without a parent dir containing a package.json file and check the external generator was used.
  - you should be asked for a description of your view
  - your current directory should have the new view in it (including dev and script directories and a whole lot more!)
  - running `npm run build-dev && open dev/index.html` should load your default browser and if you look the source you should see an element with the class of `ui-component-[your view name]`
3. run `yo ui-component:external` anywhere and check the external generator is run (see above)
4. attempting to run `yo ui-component:internal` outside of a project or in a project with no `code` dir in the root should throw an error

#### Any background context you want to provide?
The React views built by this generator are intended for use in our webapp-template based apps that we're beginning to build.

There's recently been some name changing, so webapp-views are now ui-components. This is to bring this work in line with the css toolkit the UXUI team are working on.

#### What are the relevant tickets?
[TAR-197](https://holidayextras.jira.com/browse/TAR-197)

#### Screenshots (if appropriate)

#### What gif best describes how you feel about this work?
![](http://media.giphy.com/media/YxMACiNIvfSc8/giphy.gif)
---
#### Quality checklist (for PR author to complete BEFORE code review):

#### IP or Developer Review:

#### Software Engineer or Developer review:

#### Software Engineer or project guru review:
